src/tests/MetaDataQueryTests.cpp

```diff
@@ -0,0 +1,207 @@
+class UsedProducerMockAlt : public MockProducer {
+public:
+  static RdKafka::ErrorCode ReturnErrorCode;
+  static int TimeOut;
+  static RdKafka::Producer *create(RdKafka::Conf *, std::string &) {
+    return new UsedProducerMockAlt;
+  }
+  RdKafka::ErrorCode
+  offsetsForTimes(std::vector<RdKafka::TopicPartition *> &Offsets,
+                  int UsedTimeOut) override {
+    for (auto Offset : Offsets) {
+      Offset->set_offset(RETURN_TIME_OFFSET);
+    }
+    UsedProducerMockAlt::TimeOut = UsedTimeOut;
+    return UsedProducerMockAlt::ReturnErrorCode;
+  }
+  RdKafka::ErrorCode metadata(bool, const RdKafka::Topic *,
+                              RdKafka::Metadata **MetaDataPtr, int) override {
+    *MetaDataPtr = new UsedMockMetadata;
+    return UsedProducerMockAlt::ReturnErrorCode;
+  }
+};
+int UsedProducerMockAlt::TimeOut;
+RdKafka::ErrorCode UsedProducerMockAlt::ReturnErrorCode =
+    RdKafka::ErrorCode::ERR_NO_ERROR;
+
+TEST_F(GetTopicOffsetTest, Success) {
+  UsedProducerMockAlt::ReturnErrorCode = RdKafka::ErrorCode::ERR_NO_ERROR;
+  std::vector<int> Partitions{4, 5};
+  std::vector<std::pair<int, int64_t>> ExpectedResult{{4, RETURN_TIME_OFFSET},
+                                                      {5, RETURN_TIME_OFFSET}};
+  EXPECT_EQ(KafkaW::getOffsetForTimeImpl<UsedProducerMockAlt>(
+                "Some_broker", "some_topic", Partitions,
+                std::chrono::system_clock::now(), 10ms),
+            ExpectedResult);
+}
+
+TEST_F(GetTopicOffsetTest, Failure) {
+  auto UsedTimeOut{234ms};
+  UsedProducerMockAlt::ReturnErrorCode = RdKafka::ErrorCode::ERR__BAD_MSG;
+  EXPECT_THROW(KafkaW::getOffsetForTimeImpl<UsedProducerMockAlt>(
+                   "Some_broker", "some_topic", {4},
+                   std::chrono::system_clock::now(), UsedTimeOut),
+               MetadataException);
+  EXPECT_EQ(std::chrono::duration_cast<std::chrono::milliseconds>(UsedTimeOut)
+                .count(),
+            UsedProducerMockAlt::TimeOut);
+}
+
+class GetTopicPartitionsTest : public ::testing::Test {};
+
+class TopicMockAlt {
+public:
+  static RdKafka::Topic *create(RdKafka::Handle *, const std::string &,
+                                RdKafka::Conf *, std::string &) {
+    return nullptr;
+  }
+};
+
+TEST_F(GetTopicPartitionsTest, Success) {
+  UsedProducerMockAlt::ReturnErrorCode = RdKafka::ErrorCode::ERR_NO_ERROR;
+  auto ExpectedPartitions = std::vector<int>{1, 3, 5};
+  auto ReceivedPartitions =
+      KafkaW::getPartitionsForTopicImpl<UsedProducerMockAlt, TopicMockAlt>(
+          "Some_broker", "some_topic", 10ms);
+  EXPECT_EQ(ReceivedPartitions, ExpectedPartitions);
+}
+
+TEST_F(GetTopicPartitionsTest, MetadataFailure) {
+  UsedProducerMockAlt::ReturnErrorCode = RdKafka::ErrorCode::ERR__TIMED_OUT;
+  try {
+    KafkaW::getPartitionsForTopicImpl<UsedProducerMockAlt, TopicMockAlt>(
+        "Some_broker", "some_topic", 10ms);
+    FAIL() << "Did not throw";
+  } catch (MetadataException &E) {
```

Comment:
`EXPECT_THROW(<calling code goes here>, MetadataException)`
Better?

Reply:
Feel free to try yourself but that does not work.

Reply:
Stupid GTEST and its macros!
Anyway putting a second set of `()` round the first parameter fixes it. I think overall it is better than what you have currently.
`EXPECT_THROW(
  (KafkaW::getPartitionsForTopicImpl<UsedProducerMockAlt, TopicMockAlt>(
          "Some_broker", "some_topic", 10ms)),
 MetadataException);`